_docs/ERP-config.md
# ERPNext Notes

## 1. Instalation

Llamar a Nacho

## 2. Patches

### 2.1 Allow zero in ROUND_OFF
```diff
diff --git a/erpnext/accounts/doctype/gl_entry/gl_entry.py b/erpnext/accounts/doctype/gl_entry/gl_entry.py
index 5c76799..9edd602 100644
--- a/erpnext/accounts/doctype/gl_entry/gl_entry.py
+++ b/erpnext/accounts/doctype/gl_entry/gl_entry.py
@@ -56,9 +56,9 @@ class GLEntry(Document):
                                        .format(self.voucher_type, self.voucher_no, self.account))
 
                # Zero value transaction is not allowed
-               if not (flt(self.debit) or flt(self.credit)):
-                       frappe.throw(_("{0} {1}: Either debit or credit amount is required for {2}")
-                               .format(self.voucher_type, self.voucher_no, self.account))
+               # if not (flt(self.debit) or flt(self.credit)):
+               #       frappe.throw(_("{0} {1}: Either debit or credit amount is required for {2}")
+               #               .format(self.voucher_type, self.voucher_no, self.account))
 
        def pl_must_have_cost_center(self):
                if frappe.db.get_value("Account", self.account, "report_type") == "Profit and Loss":
```

### 2.2 Get the currency of the subscription plan when generating the invoice AND get the right debit_to account
```diff
diff --git a/erpnext/accounts/doctype/subscription/subscription.py b/erpnext/accounts/doctype/subscription/subscription.py
index 2bb2c00..3dc2121 100644
--- a/erpnext/accounts/doctype/subscription/subscription.py
+++ b/erpnext/accounts/doctype/subscription/subscription.py
@@ -241,6 +241,11 @@ class Subscription(Document):
                invoice.posting_date = self.current_invoice_start
                invoice.customer = self.customer
 
+                subscription_plan = frappe.get_doc("Subscription Plan", self.plans[0].plan)
+                invoice.currency = subscription_plan.currency
+                invoice.party_account_currency = subscription_plan.currency
+                invoice.debit_to = "Debtors-%s - BL" % invoice.currency
+
                # Subscription is better suited for service items. I won't update `update_stock`
                # for that reason
                items_list = self.get_items_from_plans(self.plans, prorate)
```

### 2.3 Method 'get_pending_invoices_to_be_paid' for Subscription Invoice
`/api/method/erpnext.accounts.doctype.subscription_invoice.subscription_invoice.get_pending_invoices_to_be_paid`
```diff
diff --git a/erpnext/accounts/doctype/subscription_invoice/subscription_invoice.py b/erpnext/accounts/doctype/subscription_invoice/subscription_invoice.py
index 6f459b4..a40f312 100644
--- a/erpnext/accounts/doctype/subscription_invoice/subscription_invoice.py
+++ b/erpnext/accounts/doctype/subscription_invoice/subscription_invoice.py
@@ -3,7 +3,17 @@
 # For license information, please see license.txt
 
 from __future__ import unicode_literals
+import frappe
 from frappe.model.document import Document
 
 class SubscriptionInvoice(Document):
-       pass
+  pass
+
+@frappe.whitelist()
+def get_pending_invoices_to_be_paid():
+  pending_invoices = frappe.db.sql("""SELECT `tabSubscription Invoice`.* FROM `tabSubscription Invoice`
+  JOIN `tabSales Invoice` 
+      ON `tabSales Invoice`.name = `tabSubscription Invoice`.invoice
+          AND `tabSales Invoice`.status IN ('Unpaid', 'Overdue')
+              AND `tabSales Invoice`.docstatus=1""", as_dict=1)
+  return pending_invoices
```

### 2.4 Method 'get_delivery_note_of_sales_order' for Delivery Note
```diff
diff --git a/erpnext/stock/doctype/delivery_note/delivery_note.py b/erpnext/stock/doctype/delivery_note/delivery_note.py
index 1eb2b0985a..7264d8a748 100644
--- a/erpnext/stock/doctype/delivery_note/delivery_note.py
+++ b/erpnext/stock/doctype/delivery_note/delivery_note.py
@@ -585,3 +585,8 @@ def make_sales_return(source_name, target_doc=None):
 def update_delivery_note_status(docname, status):
        dn = frappe.get_doc("Delivery Note", docname)
        dn.update_status(status)
+
+
+@frappe.whitelist()
+def get_delivery_note_of_sales_order(sales_order):
+       return frappe.db.sql("SELECT parent FROM `tabDelivery Note Item` WHERE against_sales_order='%s'" % sales_order)
\ No newline at end of file
```

### 2.5 Method 'get_note_item_by_serial_no' for Delivery Note Item
```diff
diff --git a/erpnext/stock/doctype/delivery_note_item/delivery_note_item.py b/erpnext/stock/doctype/delivery_note_item/delivery_note_item.py
index aaca802..4027c07 100644
--- a/erpnext/stock/doctype/delivery_note_item/delivery_note_item.py
+++ b/erpnext/stock/doctype/delivery_note_item/delivery_note_item.py
@@ -10,3 +10,10 @@ from erpnext.controllers.print_settings import print_settings_for_item_table
 class DeliveryNoteItem(Document):
        def __setup__(self):
                print_settings_for_item_table(self)
+
+@frappe.whitelist()
+def get_note_item_by_serial_no(serial_no):
+  return frappe.db.sql("""SELECT note_item.item_code
+                  FROM `tabDelivery Note Item` AS note_item , `tabDelivery Note Item` AS note_item_1
+                  WHERE note_item.against_sales_order = note_item_1.against_sales_order AND note_item_1.serial_no='%s'""" % serial_no)
+
```


## 3. Operation


### 3.1 Restart
#### Hidrogeno

`sudo kill -9 $(pidof -x supervisord) && yes | sudo bench setup production operador`

#### S30

Tiene que ser root.
`cd /opt/apps/frappe-bench && kill -9 $(pidof -x supervisord) && yes | bench setup production operador`
  
## 4. Config setup

### 4.1 Custom Fields - ERP
| Resource Name   | Field Type    | Field Name             | Description                 | Required | Extra Info
---               | ---           | ---                    | ---                         | :---:    |:---:
Contact           | Data          | hhub_userid            | user id del hub             | si       |
Payment Entry     | Data          | hhub_payment_gateway   | nombre de pasarela          | si       |
Payment Entry     | Data          | hhub_payment_trxid     | id de transaccion           | si       |
Payment Entry     | Data          | hhub_card_issuerName   | issuer de la tarjeta        | si       |
Payment Entry     | Data          | hhub_card_last4        | ultimos 4 de tarjeta        | si       |
Payment Entry     | Data          | hhub_card_holder       | titular de tarjeta          | si       |
Payment Entry     | Data          | hhub_payment_user      | id de usuario en la plataforma de pagos | si |
Payment Entry     | Data          | statement_descriptor   | descripcion de pago en resumen | si |
Delivery Note     | Data          | shipping_airway_number | Airway Bill Number (DHL)    | no       | allow on submit
Sales Order Item  | Data          | subscription_code      | codigo de suscripcion       | no       |
Sales Order Item  | Data          | interpretation_code    | interpretacion ecg          | no       |
Sales Invoice     | Data          | payment_attempts       | nros de intentos de pagos   | si       | default 0 & allow on submit
Address           | Data          | address_recipient      | persona asociada al address | si       |
Subscription      | Section Break | sb_5                   | Included items              | no       |
Subscription      | Table         | items                  | Items                       | no       | Options: Subscription Item

### 4.1 Custom DocType - ERP

#### Subscription Item 
```
Name: Subscription Item
Module: Selling
Fields: 
     - Serial Number
       fieldname : serial_number
       fieldtype : Link
       label     : Serial Number
       options   : Serial No
     - Item
       fieldname : item
       fieldtype : Link
       label     : Item
       options   : Item
```

## 5 Data Setup

### 5.0 Companias

1. Bambus Life (Global)
   Ventas web (en todos los paises)
2. Tingelmar (UY)

Proveedores
1. CardrioSecur

Clientes
1. SEMM

### 5.1 Productos

| | | | 
|-|-|-|
CS-ACT-USBC  | CardioSecur Active (usb-c) 
CS-ACT-MUSB  | CardioSecur Active (m-usb)
CS-ACT-LIGHT | CardioSecur Active (lightning)
CS-PRO-LIGHT | CardioSecur Pro (lightning)
CS-ACT-SUBS    | CardioSecur Active Subscription
CS-PRO-SUBS-12 | CardioSecur Pro Subscription 12L
CS-PRO-SUBS-22 | CardioSecur Pro Subscription 22L
CS-ECG-INTER   | Interpretación de Electrocardiograma

### 5.2 Warehouses
Default Warehouse - Stores BL
  
### 5.3 Price Lists

*CardioSecur Prices (buying)*  
  
| | | | 
|-|-:|-|
CS-ACT-USBC    | 58,23 | USD
CS-ACT-MUSB    | 58,23 | USD
CS-ACT-LIGHT   | 58,23 | USD
CS-PRO-LIGHT   | 91,98 | EUR


*Precios Web*  
  
| | | | 
|-|-:|-|
CS-ACT-USBC    | 120,00 | USD
CS-ACT-MUSB    | 120,00 | USD
CS-ACT-LIGHT   | 120,00 | USD
CS-PRO-LIGHT   | 200,00 | USD
CS-ACT-SUBS    |  19,90 | USD
CS-PRO-SUBS-12 |  29,90 | USD
CS-PRO-SUBS-22 |  39,90 | USD
CS-ECG-INTER   |  29,00 | USD


CHANGELOG.md

### FALTA
 + PageSpeed
   + Add PurgedCSS
   + Agregar descriptions!!
   + Agregar structured data: 
     + Corporate Contact: https://developers.google.com/search/docs/data-types/corporate-contact (FALTA TELEFONO!!)
     + FAQ: https://developers.google.com/search/docs/data-types/faqpage
     + Product: https://developers.google.com/search/docs/data-types/product
     + Bradcrumb: https://developers.google.com/search/docs/data-types/breadcrumb

### TEST
#### -------- COMPRAS EN UY
1. Compra Active UY (global)
   User   : arondella+pacuy04111903@tingelmar.com
   Card   : 5161 4413 1585 2061
   A. Delivery (creacion de device y subscripcion)
   B. Cobro de mensualidad (cuota 1)

2. Compra Active UY (local)
   User    : arondella+pacuy04111904@tingelmar.com
   A. Delivery (creacion de device y subscripcion)
   B. Cobro de mensualidad (cuota 1)

3. Registro de medico
   User   : arondella+doc@tingelmar.com

4. Compra Pro UY
   User   : arondella+docpro@tingelmar.com
   Device : MM5092
   Organizacion : 61
   Paciente 1   : 871

5. Compra Pro UY (2)
   User   : arondella+docpro2@tingelmar.com
   Device : MM5091
   Organizacion : X
   Paciente 1   : X

#### -------- COMPRAS EN CL 
1. Compra Active CL (global)
   User   : arondella+paccl04111903@tingelmar.com
   Card   : 5031 7557 3453 0604 (123 - 11/25)
   A. Delivery (creacion de device y subscripcion)
   B. Cobro de mensualidad (cuota 1)
     19,97 + 3,79 = 23,76 ( muestra: 23,77 )

1. Compra Active CL (local)
   User    : arondella+paccl04111904@tingelmar.com
   A. Delivery (creacion de device y subscripcion)
   B. Cobro de mensualidad (cuota 1)


https://api.mercadopago.com/v1/payments/22489289?access_token=TEST-4450788232191490-042612-40ccc8aacefa128d70ac5e97783b8b78-429640315

#### -------- PLAN FAMILIAR
 OwnerId = 540 (PiD: 1089) arondella+paccl05111901@tingelmar.com
 Invited = 535 (PiD: 1084) arondella+paccl04111901@tingelmar.com
 appointmentDevicesId : 390

### POST VENTA
  user arondella+enzo@tingelmar.com
  http://localhost:8000/uy/compra/J2YxVMRDaNTmLLQklxnQ/-



## 0.9.11 - BETA - 2021-09-24
### Fix
 - Se ajusta banner y landing de casmu

## 0.9.10 - BETA - 2021-09-24
### New
 - Se cambia el banner del home

## 0.9.9 - BETA - 2021-09-23
### New
 - Se agrega página landing de sorteo de casmu

## 0.9.8 - BETA - 2021-09-23
### New
 - Se cambia titulo de home
 
## 0.9.7 - BETA - 2021-05-03
### New
 - Se publica seccion de prensa

## 0.9.6 - BETA - 2021-03-11
### New
 - Se ajusta seccion de prensa

## 0.9.5 - BETA - 2021-03-01
### New
 - Se agrega seccion de prensa

## 0.9.4 - BETA - 2020-09-28
### New
 - Se vuelve al banner de Covid.

## 0.9.3 - BETA - 2020-09-25
### New
 - Se modifica para mostrar los precios cuando no hay stock en CSActive y CSPro.

## 0.9.2 - BETA - 2020-09-21
### New
 - Se ajusta logo en metas de favicon y se agrega manifest.
 - Se agregan tags de OpenGraph

## 0.9.1 - BETA - 2020-09-16
### New
 - Se agrega descarga de brochures
 - Se agrega registro de paciente
 - Se agrega opcion de registro de pacientes y medicos en el menu principal

### Fixes
 - Se corrige zona horaria de citas en perfil

## 0.9.0 - BETA - 2020-09-07
### New
 - Se agrega nueva pagina para completar el registro de nuevo usuario paciente.

## 0.8.4 - BETA - 2020-08-31
### Fixes
 - Se corrige envio de mail en reserva

## 0.8.3 - BETA - 2020-08-26
### New
 - Se agrega reCaptcha a formulario de registro de medicos. 

## 0.8.2 - BETA - 2020-08-24
### Fixes
 - Se ajusta mostrado de error en registro de medicos

## 0.8.1 - BETA - 2020-08-13
### Fixes
 - Se ajustan fechas en listados en perfil
 - Se agrega `showSetNickname` en control de video en pagina de cita.
 - Se ajustan botones de volver en wizard de reservas

## 0.8.0 - BETA - 2020-08-05
### New
 - Se agrega reserva de cita con soporte de video

## 0.7.2 - BETA - 2020-06-25
### New
 - Se quita el precio de lsa derivaciones en pagina de order de pro. 

## 0.7.1 - BETA - 2020-05-25
### Fixes
 - Se corrige link a pagina de accesorios

## 0.7.0 - BETA - 2020-05-22
### New
 - Nuevo diseño de paginas home, entidades medicas.
 - Se agrega pagina de Covid y portal médico

## 0.6.1 - BETA - 2020-05-07
### New
 - Se agrega pagina de accesorios
 - Compra de informe de diagnostico
 - Ajuste de calculo de subtotal:
   - OrderView - calculo de impuestos y subtotal
   - CartView  - calculo de impuestos y subtotal

src/Utils/Settings.php
<?php
namespace App\Utils;

use Symfony\Component\HttpFoundation\Request;

// Third Party Payment
use MercadoPago;

class Settings {

  const countries = array (
    'ar' => array('code' => 'AR', 'name' => 'Argentina'),
    'bo' => array('code' => 'BO', 'name' => 'Bolivia'),
    'br' => array('code' => 'BR', 'name' => 'Brasil'),
    'cl' => array('code' => 'CL', 'name' => 'Chile'),
    'co' => array('code' => 'CO', 'name' => 'Colombia'),
    'cr' => array('code' => 'CR', 'name' => 'Costa Rica'),
    'cu' => array('code' => 'CU', 'name' => 'Cuba'),
    'ec' => array('code' => 'EC', 'name' => 'Ecuador'),
    'gt' => array('code' => 'GT', 'name' => 'Guatemala'),
    'sv' => array('code' => 'SV', 'name' => 'El Salvador'),
    'hn' => array('code' => 'HN', 'name' => 'Honduras'),
    'mx' => array('code' => 'MX', 'name' => 'Mexico'),
    'ni' => array('code' => 'NI', 'name' => 'Nicaragua'),
    'py' => array('code' => 'PY', 'name' => 'Paraguay'),
    'pe' => array('code' => 'PE', 'name' => 'Perú'),
    'pa' => array('code' => 'PA', 'name' => 'Panamá'),
    'pr' => array('code' => 'PR', 'name' => 'Puerto Rico'),
    'do' => array('code' => 'DO', 'name' => 'República Dominicana'),
    'uy' => array('code' => 'UY', 'name' => 'Uruguay'),
    've' => array('code' => 'VE', 'name' => 'Venezuela'),
  );

  const currency_country = array(
    'AR' => 'UYU',
    //'CL' => 'UYU', // 'CLP',
    'CL' => 'CLP',
    'PY' => 'UYU',
    'BO' => 'UYU',
    'PE' => 'UYU',
    'EC' => 'UYU',
    'CO' => 'UYU',
    'VE' => 'UYU',
    'PA' => 'UYU',
    'CR' => 'UYU',
    'NI' => 'UYU',
    'SV' => 'UYU',
    'HN' => 'UYU',
    'GT' => 'UYU',
    'MX' => 'UYU',
    'PR' => 'UYU',
    'DO' => 'UYU',
    'CU' => 'UYU',
    'BR' => 'UYU',
    'UY' => 'UYU'
  );

  const currency_name = array(
    'CLP' => 'pesos chilenos',
    'UYU' => 'pesos uruguayos',
    'USD' => 'dolares americanos'
  );

  // const mp_keys = array(
  //   'UY' => array(
  //     'private' => 'TEST-3945512895144203-022509-d3458a75fa3768bbe00cbda3b9790ffb-407141024',
  //     'public'  => 'TEST-ca3cbeec-5d8c-4d29-8df1-59b116b16afc'
  //   ),
  //   'CL' => array(
  //     // 'private' => 'TEST-4450788232191490-042612-40ccc8aacefa128d70ac5e97783b8b78-429640315',
  //     // 'public'  => 'TEST-100845d7-0ced-4802-afef-43ea4a5b545f'
  //     'private' => 'TEST-3945512895144203-022509-d3458a75fa3768bbe00cbda3b9790ffb-407141024',
  //     'public'  => 'TEST-ca3cbeec-5d8c-4d29-8df1-59b116b16afc'
  //   )
  // );

  const mp_payment_methods = array(
    'DEFAULT' => array(
      array(
        "css_class" => "visa",
        "name" => "Visa"
      ),
      array(
        "css_class" => "amex",
        "name" => "American Express"
      ),
      array(
        "css_class" => "master",
        "name" => "Mastercard"
      )
    ),
    'UY' => array(
      // array(
      //   'css_class' => 'account_money',
      //   'name' => 'Dinero en cuenta Mercado Pago',
      // ),
      array(
        'css_class' => 'visa',
        'name' => 'Visa'
      ),
      array(
        'css_class' => 'master',
        'name' => 'Mastercard'
      ),
      array(
        'css_class' => 'diners',
        'name' => 'Diners'
      ),
      // array(
      //   'css_class' => 'abitab',
      //   'name' => 'Abitab'
      // ),
      array(
        'css_class' => 'oca',
        'name' => 'Oca'
      ),
      // array(
      //   'css_class' => 'redpagos',
      //   'name' => 'Red Pagos'
      // ),
      array(
        'css_class' => 'lider',
        'name' => 'Lider'
      ),
      array(
        'css_class' => 'debvisa',
        'name' => 'Visa Débito'
      )
    ),
    'CL' => array (
      array(
        "css_class" => "visa",
        "name" => "Visa"
      ),
      array(
        "css_class" => "amex",
        "name" => "American Express"
      ),
      array(
        "css_class" => "master",
        "name" => "Mastercard"
      ),
      array(
        'css_class' => 'diners',
        'name' => 'Diners'
      )
    )
  );

  const mp_script_src = array(
    'UY' => 'https://www.mercadopago.com.uy/integrations/v1/web-tokenize-checkout.js',
    'CL' => 'https://www.mercadopago.cl/integrations/v1/web-tokenize-checkout.js'
    //'CL' => 'https://www.mercadopago.com.uy/integrations/v1/web-tokenize-checkout.js'
  );

  const mp_status_errors = array (
    // Rejected
    "cc_rejected_bad_filled_card_number"   => "Revisa el número de tarjeta.",
    "cc_rejected_bad_filled_date"          => "Revisa la fecha de vencimiento.",
    "cc_rejected_bad_filled_other"         => "Revisa los datos.",
    "cc_rejected_bad_filled_security_code" => "Revisa el código de seguridad.",
    "cc_rejected_blacklist"                => "No pudimos procesar tu pago.",
    "cc_rejected_call_for_authorize"       => "Debes autorizar ante el pago a Mercado Pago",
    "cc_rejected_card_disabled"            => "Llama al emisor/banco para que active tu tarjeta. El teléfono está al dorso de tu tarjeta.",
    "cc_rejected_card_error"               => "No pudimos procesar tu pago.",
    "cc_rejected_duplicated_payment"       => "Ya hiciste un pago por ese valor.",
    "cc_rejected_high_risk"                => "Tu pago fue rechazado. Elige otro de los medios de pago.",
    "cc_rejected_insufficient_amount"      => "Tu tarjeta no tiene fondos suficientes.",
    "cc_rejected_invalid_installments"     => "Tu tarjeta no procesa pagos en las cuotas especificadas.",
    "cc_rejected_max_attempts"             => "Llegaste al límite de intentos permitidos. Elige otra tarjeta.",
    "cc_rejected_other_reason"             => "El emisor/banco no procesó el pago.",
    // in_process
    "pending_contingency"                  => "Estamos procesando el pago. En menos de 2 días hábiles te enviaremos por e-mail el resultado.",
    "pending_review_manual"                => "Estamos procesando el pago. En menos de 2 días hábiles te diremos por e-mail si se acreditó o si necesitamos más información.",
    // aprroved
    "accredited"                           => "Listo, se acreditó tu pago!"
  );

  const support_parties = array (
    'uy'              => 'sacuy',
    'cl'              => 'saccl',
    'default'         => 'sac',
    'uy_logged'       => 'gsouy',
    'cl_logged'       => 'gsocl',
    'default_logged'  => 'gso'
  );

  const SERVICE_PRODUCT_DIAG_REP_ECG = 'DIAG_REP_ECG';
  const SERVICE_PRODUCT_ENCOUNTER    = 'ENCOUNTER';

  public static function getMPPaymentMethods($countryCode){
    if (array_key_exists($countryCode, self::mp_payment_methods)){
      return self::mp_payment_methods[$countryCode];
    }else{
      return self::mp_payment_methods['DEFAULT'];
    }
  }

  public static function getMPKeys($countryCode){
    // Build Country dedicated MP Keys
    $ENV_PRIVATE_KEY   = "BL_MP".$countryCode."_PRIVATE_KEY";
    $ENV_PUBLIC_KEY    = "BL_MP".$countryCode."_PUBLIC_KEY";
    $ENV_CLIENT_ID     = "BL_MP".$countryCode."_CLIENT_ID";
    $ENV_CLIENT_SECRET = "BL_MP".$countryCode."_CLIENT_SECRET";
    // Build default keys (UY)
    $MPKEYS = array(
      'PRIVATE_KEY'   => $_SERVER['BL_MPUY_PRIVATE_KEY'],
      'PUBLIC_KEY'    => $_SERVER['BL_MPUY_PUBLIC_KEY'],
      'CLIENT_ID'     => $_SERVER['BL_MPUY_CLIENT_ID'],
      'CLIENT_SECRET' => $_SERVER['BL_MPUY_CLIENT_SECRET']
    );
    // Override if exists
    if (array_key_exists($ENV_PRIVATE_KEY, $_SERVER)){
      $MPKEYS['PRIVATE_KEY'] = $_SERVER[$ENV_PRIVATE_KEY];
    }
    if (array_key_exists($ENV_PUBLIC_KEY, $_SERVER)){
      $MPKEYS['PUBLIC_KEY'] = $_SERVER[$ENV_PUBLIC_KEY];
    }
    if (array_key_exists($ENV_CLIENT_ID, $_SERVER)){
      $MPKEYS['CLIENT_ID'] = $_SERVER[$ENV_CLIENT_ID];
    }
    if (array_key_exists($ENV_CLIENT_SECRET, $_SERVER)){
      $MPKEYS['CLIENT_SECRET'] = $_SERVER[$ENV_CLIENT_SECRET];
    }
    return $MPKEYS;
  }

  public static function get_currency_by_country($country) {
    if ($country == NULL || !strlen($country)) {
      $currency_code = 'USD';
    } else {
      $currency_code = self::currency_country[strtoupper($country)];
    }

    return Array(
      'code' => $currency_code,
      'name' => self::currency_name[$currency_code]
    );
  }

  public static function get_country_from_locale(Request $request){
    $locale = strtolower($request->getLocale());
    return Self::countries[$locale];
  }

  public static function resolveCountry($user, Request $request){
    // $countryCode = NULL;
    // if (isset($user)){
    //   if (array_key_exists('countryCode', $user)){
    //     if (strlen($user['countryCode']) > 0 ){
    //       $countryCode = strtolower($user['countryCode']);
    //     }
    //   }
    // }
    // if ($countryCode == NULL){
      $countryCode = strtolower($request->getLocale());
    // }
    return Self::countries[$countryCode];
  }

  public static function convertMilitaryTime($mTime, $ampm=FALSE){
    $hours = intval($mTime / 100);
    $mins  = intval($mTime % 100);
    $strTime = '';
    if ($ampm){
      $strAMPM = 'AM';
      if ($hours < 10){
       $strTime .= "0" . $hours;
      }else{
        if ($hours > 12){
          $strTime .= ($hours-12);
          $strAMPM = 'PM';
        }else{
          $strTime .= $hours;
        }
      }
      if ($mins < 10){
        $strTime .= ":0" . $mins;
      }else{
        $strTime .= ":" . $mins;
      }
      $strTime .= " " . $strAMPM;
    }else{
      if ($hours < 10){
        $strTime .= "0" . $hours;
      }else{
        $strTime .= $hours;
      }
      if ($mins < 10){
        $strTime .= ":0" . $mins;
      }else{
        $strTime .= ":" . $mins;
      }
    }
    return $strTime;
  }


  public static function convertStatus($status){
    if ($status == 'BOOKED'){
      return 'Agendada';
    }else if ($status == 'CANCELLED'){
      return 'Cancelada';
    }else{
      return $status;
    }
  }

  //
  // Fast Payment processor (TEST)
  //
  public static function doFastPay(Request $request, $token, $logger, $authCtrl, $payInfo){
    //$this->logger->info(sprintf("Default|doFastPay| Recovering session data with id %s...", $sid));
    // Ensure sessionId from previous checkout step
    // $session = $this->get('session');
    // $session->setId($sid);
    // $session->start();

    // Get Payment details
    $user        = $authCtrl->getUserData($request);
    $country     = Settings::resolveCountry($user, $request);
    $navCountry  = strtoupper($request->getLocale());

    $logger->info('Default|doFastPay|1| Processing payment ***'.json_encode($payInfo).'***...');

    //
    // Setup MP
    //
    $mpKeys = Settings::getMPKeys($country['code']);
    MercadoPago\SDK::setClientId($mpKeys['CLIENT_ID']);
    MercadoPago\SDK::setClientSecret($mpKeys['CLIENT_SECRET']);
    MercadoPago\SDK::setAccessToken($mpKeys["PRIVATE_KEY"]);

    $logger->info(sprintf("Default|doFastPay|1| Start processing fast payment for %s...", $user["email"]));
    //
    // Prepare Invoice
    //
    // $logger->info("Default|doFastPay|1| Preparing invoice...");
    // $pareparedInvoice = NULL;
    // TODO
    // FOR NOW, INVOICE IS CREATED OUTSIDE THIS FUNCTION, PREVIOUS TO THIS CALL

    //
    // Get MP data
    //
    // $token             = $_REQUEST["mptoken"];
    //$payment_method_id = $_REQUEST["mppayment_method_id"];
    if (array_key_exists('installments', $_REQUEST)){
      $installments    = $_REQUEST["mpinstallments"];
    }else{
      $installments    = 1;
    }
    //
    // Resolve MP customer
    //
    $logger->info(sprintf("Default|doFastPay|2| Searching customer %s", $user["email"]));
    $existing_customer = MercadoPago\Customer::search(array("email" => $user['email']));
    if (!count($existing_customer)) {
      $customer = new MercadoPago\Customer();
      $customer->email = $user["email"];
      $this->logger->info(sprintf("Default|doFastPay|2| Saving MP customer with email: %s", $user["email"]));
      $resClient = $customer->save();
      $this->logger->debug(sprintf("Default|doFastPay|2| Customer save returned: [%s]", json_encode($resClient)));
      if ($customer->id == NULL || strlen($customer->id) == 0){
        $errMsg = "<strong>El pago no pudo ser completado</strong><br>Error en el registro de cliente";
        return Settings::cancelFP($logger, NULL, $errMsg);
      }
    } else {
      $logger->info(sprintf("Default|doFastPay|2| Using existing MP customer with email: %s", $user["email"]));
      $customer = $existing_customer[0];
    }
    // ---------------------------------
    // DO PAYMENT IN MP
    // ---------------------------------
    $payment = new MercadoPago\Payment();
    $payment->transaction_amount = $payInfo['amountToPay'];
    $payment->token              = $token;
    if ($installments){
      $payment->installments       = $installments;
    }
    //$payment->payment_method_id  = $payment_method_id;
    //$payment->issuer_id          = $issuer_id;
    $payment->binary_mode        = TRUE;
    
    $payment->payer = array(
      "id" => $customer->id
    );
    $logger->info(sprintf("Default|doFastPay|2| Requesting payment at MP with token: %s, amount: %s, customerId: %s", $token, $payInfo['amountToPay'], $customer->id));
    $logger->debug('----------------------------------------------------');
    $logger->debug(print_r($payment, TRUE));
    $logger->debug('----------------------------------------------------');
    $payment->save();
    if (!isset($payment->card) || (isset($payment->card) && $payment->card->id == NULL)) {
      $logger->info("Default|doFastPay|2| Saving card at MP...");
      $card = new MercadoPago\Card();
      $card->token = $token;
      $card->customer_id = $customer->id;
      $savedCard = $card->save();
      $logger->debug(sprintf("Default|doFastPay|2| Card save returned: [%s]", print_r($savedCard)));
      if ($card->id == NULL || strlen($card->id) == 0){
        $errMsg = "<strong>El pago no pudo ser completado</strong><br>Error en el registro de tarjeta";
        $logger->warning(sprintf("Default|doFastPay|2| Error saving card (card id empty)!!"));
        return Settings::cancelFP($logger, NULL, $errMsg);
      }
    } else {
      $logger->info(sprintf("Default|doFastPay|2| Searching card id %s...", $payment->card->id));
      $card = MercadoPago\Card::find_by_id($payment->card->id);
      if (!isset($card->issuer)) {
        $logger->info(sprintf("Default|doFastPay|2| Card id %s loaded.", $payment->card->id));
        $card = $payment->card;
      }else{
        $logger->warning(sprintf("Default|doFastPay|2| Card id %s is missing issuer!", $payment->card->id));
      }
    }
    //
    // CHECK PAYMENT RESULT
    //

    // TEST PAYMENT SIMULATED - BEGIN
    // $payment = new \stdClass();
    // $payment->id = 'abcdefghijklmnopqrstuvxyz';
    // $payment->statement_descriptor = 'statement_descriptor';
    // $payment->status = 'approved';
    // $payment->card = new \stdClass();
    // $payment->card->last_four_digits = '0000';
    // $payment->card->cardholder = new \stdClass();
    // $payment->card->cardholder->name = 'TITULAR TEST';
    // $card = new \stdClass();
    // $card->issuer = new \stdClass();
    // $card->issuer->name = 'visa';
    // TEST PAYMENT SIMULATED - END

    if ($payment->status == 'approved' || $payment->status == 'in_process'){
      //
      // PAYMENT APPROVED
      //
      $logger->info(sprintf("Default|doFastPay|2|MP payment was approved with id: %s.", $payment->id));
      
      $paymentGW = array(
        'gateway'              => "MERCADOPAGO",
        'transaction_id'       => $payment->id,
        'user_id'              => $customer->id,
        'statement_descriptor' => $payment->statement_descriptor,
        'external_userid'      => $customer->id
      );
      if ($payment->status == 'approved'){
        $paymentGW['status'] = 'APPROVED';
      }else{
        $paymentGW['status'] = 'IN_PROGRESS';
      }
      $paymentGW['issuer_name'] = 'Tarjeta';
      if (isset($card->issuer)) {
        $paymentGW['issuer_name'] = $card->issuer->name;
      }
      if (isset($card->last_four_digits)){
        $paymentGW['card_last_four'] = $card->last_four_digits;
      }else if (isset($payment->card->last_four_digits)){
        $paymentGW['card_last_four'] = $payment->card->last_four_digits;
      }
      if (isset($card->cardholder)){
        $paymentGW['card_holder'] = $card->cardholder->name;
      }else if (isset($payment->card->cardholder)){
        $paymentGW['card_holder'] = $payment->card->cardholder->name;
      }

      //
      // STEP 3 | Return payment info in order to confirm docs
      //
      return $paymentGW;
    }else{
      //
      // PAYMENT FAILED
      //
      $this->logger->error("Default|doFastPay|2| Error on PASARELLA, payment was not completed! --> ". json_encode($paymentDetails));
      $this->logger->error(sprintf("Default|doFastPay|2| Payment was ***%s***", json_encode($payment)));
      // 
      // New handler
      //
      $errMsg = "<strong>El pago no pudo ser completado</strong><br>";
      if (array_key_exists($payment->status_detail, Settings::mp_status_errors)){
        $errMsg .= Settings::mp_status_errors[$payment->status_detail];
      }
      return Settings::cancelFP($logger, NULL, $errMsg);
    }
  }
  private static function cancelFP($logger, $preparedInvoice, $errMsg) {
    $logger->info(sprintf("Default|doFastPay|2|Payment could NOT BE COMPLETED! All ERP documents will be deleted."));

    // Cancel Order
    // $this->logger->info(sprintf("CheckoutCtrl|doFastPay|2|Canceling order Intl %s - Local %s...", $pareparedPurchase['orderIntl']['name'], $pareparedPurchase['orderLocal']['name']));
    // $cancelOrderRes = $this->hubClient->cancelOrder($pareparedPurchase['orderIntl']['name'], $pareparedPurchase['orderLocal']['name']);
    // if (!$cancelOrderRes['resolved'] || $cancelOrderRes['status_code'] != '200'){
    //   // Order could not be cancelled
    //   // Send email to operations 
    //   $this->logger->error("CheckoutCtrl|processPayment|2|Error cancelling order (".$cancelOrderRes['status_code']." | " . json_encode($cancelOrderRes['data']) . ")");
    //   // ------------------------------
    //   // TODO!!!!!!!!!!!!!!!!!!!!!!!!!!
    //   // ------------------------------
    //   // SEND MAIL TO OPERATIONS
    // }else{
    //   $this->logger->info(sprintf("CheckoutCtrl|processPayment|2|Order cancelled successfully."));
    // }
    
    return array(
      'status' => 'error',
      'errMsg' => $errMsg
    );
  }

};

?>